File: apps/api-rs/src/models/agent.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use tokio::sync::mpsc::UnboundedReceiver;
use uuid::Uuid;

use super::{Confidence, Usage};
use crate::models::citation::CitationAnchor;
use crate::models::trace::{ResolvedRef, RetrievalPlan, RetrievalTrace};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum AgentMode {
    Answerer,
    Clarifier,
    Summarizer,
    Comparer,
    Analyst,
    Navigator,
    Reviewer,
}

impl std::fmt::Display for AgentMode {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            AgentMode::Answerer => write!(f, "answerer"),
            AgentMode::Clarifier => write!(f, "clarifier"),
            AgentMode::Summarizer => write!(f, "summarizer"),
            AgentMode::Comparer => write!(f, "comparer"),
            AgentMode::Analyst => write!(f, "analyst"),
            AgentMode::Navigator => write!(f, "navigator"),
            AgentMode::Reviewer => write!(f, "reviewer"),
        }
    }
}

impl std::str::FromStr for AgentMode {
    type Err = String;
    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "answerer" => Ok(AgentMode::Answerer),
            "clarifier" => Ok(AgentMode::Clarifier),
            "summarizer" => Ok(AgentMode::Summarizer),
            "comparer" => Ok(AgentMode::Comparer),
            "analyst" => Ok(AgentMode::Analyst),
            "navigator" => Ok(AgentMode::Navigator),
            "reviewer" => Ok(AgentMode::Reviewer),
            _ => Err(format!("unknown agent mode: {s}")),
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ConversationTurn {
    pub user_message: String,
    pub assistant_answer: String,
    #[serde(default)]
    pub citations: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentRequest {
    pub tenant_id: Uuid,
    pub user_id: Uuid,
    pub conversation_id: Uuid,
    pub user_message_id: Uuid,
    pub assistant_message_id: Uuid,
    pub original_query: String,
    pub effective_kb_ids: Vec<Uuid>,
    pub history: Vec<ConversationTurn>,
    pub options: AgentOptions,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentOptions {
    #[serde(default)]
    pub mode: Option<AgentMode>,
    #[serde(default = "default_tone")]
    pub tone: String,
    #[serde(default = "default_true")]
    pub proactive_followup: bool,
    #[serde(default = "default_max_followup")]
    pub max_followup_suggestions: usize,
    #[serde(default = "default_true")]
    pub allow_analyst_mode: bool,
    #[serde(default = "default_true")]
    pub require_citation: bool,
    #[serde(default)]
    pub generation: GenerationConfig,
    #[serde(default)]
    pub retrieval: RetrievalRuntimeConfig,
    #[serde(default)]
    pub runtime: AgentRuntimeConfig,
}

fn default_tone() -> String {
    "concise_warm".to_string()
}
fn default_true() -> bool {
    true
}
fn default_max_followup() -> usize {
    2
}

impl Default for AgentOptions {
    fn default() -> Self {
        Self {
            mode: None,
            tone: default_tone(),
            proactive_followup: true,
            max_followup_suggestions: default_max_followup(),
            allow_analyst_mode: true,
            require_citation: true,
            generation: GenerationConfig::default(),
            retrieval: RetrievalRuntimeConfig::default(),
            runtime: AgentRuntimeConfig::default(),
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentRuntimeConfig {
    #[serde(default = "default_true")]
    pub hyde_enabled: bool,
    #[serde(default = "default_max_react_steps")]
    pub max_react_steps: usize,
    #[serde(default = "default_max_queries_per_step")]
    pub max_queries_per_step: usize,
    #[serde(default = "default_max_history_turns")]
    pub max_history_turns: usize,
    #[serde(default = "default_max_history_chars")]
    pub max_history_chars: usize,
    #[serde(default = "default_max_context_chars")]
    pub max_context_chars: usize,
    #[serde(default = "default_max_repair_attempts")]
    pub max_repair_attempts: usize,
}

fn default_max_react_steps() -> usize {
    4
}
fn default_max_queries_per_step() -> usize {
    4
}
fn default_max_history_turns() -> usize {
    12
}
fn default_max_history_chars() -> usize {
    24_000
}
fn default_max_context_chars() -> usize {
    30_000
}
fn default_max_repair_attempts() -> usize {
    3
}

impl Default for AgentRuntimeConfig {
    fn default() -> Self {
        Self {
            hyde_enabled: true,
            max_react_steps: default_max_react_steps(),
            max_queries_per_step: default_max_queries_per_step(),
            max_history_turns: default_max_history_turns(),
            max_history_chars: default_max_history_chars(),
            max_context_chars: default_max_context_chars(),
            max_repair_attempts: default_max_repair_attempts(),
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GenerationConfig {
    #[serde(default = "default_model")]
    pub model: String,
    #[serde(default = "default_temperature")]
    pub temperature: f64,
    #[serde(default = "default_max_output_tokens")]
    pub max_output_tokens: u32,
}

fn default_model() -> String {
    "qwen-turbo".to_string()
}
fn default_temperature() -> f64 {
    0.2
}
fn default_max_output_tokens() -> u32 {
    1200
}

impl Default for GenerationConfig {
    fn default() -> Self {
        Self {
            model: default_model(),
            temperature: default_temperature(),
            max_output_tokens: default_max_output_tokens(),
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RetrievalRuntimeConfig {
    #[serde(default = "default_dense_top_k")]
    pub dense_top_k: usize,
    #[serde(default = "default_bm25_top_k")]
    pub bm25_top_k: usize,
    #[serde(default = "default_rrf_top_k")]
    pub rrf_top_k: usize,
    #[serde(default = "default_rerank_top_k")]
    pub rerank_top_k: usize,
    #[serde(default = "default_true")]
    pub rerank_enabled: bool,
}

fn default_dense_top_k() -> usize {
    100
}
fn default_bm25_top_k() -> usize {
    100
}
fn default_rrf_top_k() -> usize {
    20
}
fn default_rerank_top_k() -> usize {
    5
}
impl Default for RetrievalRuntimeConfig {
    fn default() -> Self {
        Self {
            dense_top_k: default_dense_top_k(),
            bm25_top_k: default_bm25_top_k(),
            rrf_top_k: default_rrf_top_k(),
            rerank_top_k: default_rerank_top_k(),
            rerank_enabled: true,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CitationOutput {
    pub index: i32,
    pub chunk_id: Uuid,
    pub doc_id: Uuid,
    pub doc_title: String,
    pub page_range: Vec<i32>,
    pub quote: String,
    pub score: f64,
    #[serde(default = "default_citation_source_status")]
    pub source_status: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub anchor: Option<CitationAnchor>,
}

fn default_citation_source_status() -> String {
    "available".to_string()
}

#[derive(Debug, Clone)]
pub enum AnswerStreamItem {
    Delta {
        text: String,
    },
    Citation {
        citation: CitationOutput,
    },
    Completed {
        confidence: Confidence,
        usage: Option<Usage>,
    },
    Failed {
        code: String,
        message: String,
    },
}

#[derive(Debug)]
pub struct AgentRun {
    pub assistant_message_id: Uuid,
    pub mode: AgentMode,
    pub rewritten_query: Option<String>,
    pub retrieval_plan: RetrievalPlan,
    pub retrieval_traces: Vec<RetrievalTrace>,
    pub answer_stream: UnboundedReceiver<AnswerStreamItem>,
    pub trace: AgentTrace,
    pub no_answer_reason: Option<super::NoAnswerReason>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentTrace {
    pub mode: AgentMode,
    pub mode_reason: String,
    pub rewritten_query: Option<String>,
    pub keywords: Vec<String>,
    pub resolved_refs: Vec<ResolvedRef>,
    pub retrieval_plan: RetrievalPlan,
    pub prompt_versions: PromptVersions,
    pub model: String,
    pub usage: Option<Usage>,
    pub started_at: DateTime<Utc>,
    #[serde(default)]
    pub memory_summary: String,
    #[serde(default)]
    pub react_steps: Vec<ReactStepTrace>,
    #[serde(default)]
    pub stop_reason: String,
    #[serde(default)]
    pub runtime_components: RuntimeComponents,
    #[serde(default)]
    pub cache_key: Option<String>,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct RuntimeComponents {
    #[serde(default)]
    pub reasoner: String,
    #[serde(default)]
    pub retriever: String,
    #[serde(default)]
    pub reranker: String,
    #[serde(default)]
    pub verifier: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ReactStepTrace {
    pub step: usize,
    pub action: String,
    pub decision_summary: String,
    #[serde(default)]
    pub queries: Vec<String>,
    #[serde(default)]
    pub rerank_query: Option<String>,
    #[serde(default)]
    pub hypothetical_answer: Option<String>,
    #[serde(default)]
    pub retrieved_chunk_ids: Vec<Uuid>,
    #[serde(default)]
    pub accepted_chunk_ids: Vec<Uuid>,
    #[serde(default)]
    pub warnings: Vec<String>,
    pub started_at: DateTime<Utc>,
    pub completed_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PromptVersions {
    pub persona: String,
    pub guardrail: String,
    pub mode: String,
    pub task: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RewriteOutput {
    pub rewritten_query: String,
    pub keywords: Vec<String>,
    pub hypothetical_answer: Option<String>,
    pub resolved_refs: Vec<ResolvedRef>,
    pub added_constraints: Vec<String>,
    pub removed_constraints: Vec<String>,
    pub needs_clarification: bool,
    pub clarification_question: Option<String>,
}

File: apps/api-bun/src/models/agent.ts
// 移植自 apps/api-rs/src/models/agent.rs
import type { CitationAnchor } from './citation.ts';
import type { Confidence, Usage } from './index.ts';
import type { ResolvedRef, RetrievalPlan, RetrievalTrace } from './trace.ts';

export type AgentMode =
  | 'answerer' | 'clarifier' | 'summarizer' | 'comparer' | 'analyst' | 'navigator' | 'reviewer';
export const agentModes: AgentMode[] = [
  'answerer', 'clarifier', 'summarizer', 'comparer', 'analyst', 'navigator', 'reviewer',
];
export function isAgentMode(value: string): value is AgentMode {
  return agentModes.includes(value as AgentMode);
}

export interface ConversationTurn {
  user_message: string; assistant_answer: string; citations?: string[];
}

export interface AgentRequest {
  tenant_id: string; user_id: string; conversation_id: string;
  user_message_id: string; assistant_message_id: string;
  original_query: string; effective_kb_ids: string[];
  history: ConversationTurn[]; options: AgentOptions;
}

export interface AgentOptions {
  mode?: AgentMode | null; tone: string; proactive_followup: boolean;
  max_followup_suggestions: number; allow_analyst_mode: boolean; require_citation: boolean;
  generation: GenerationOptions; retrieval: RetrievalRuntimeOptions; runtime: AgentRuntimeOptions;
}
export function defaultAgentOptions(): AgentOptions {
  return {
    mode: null, tone: 'concise_warm', proactive_followup: true, max_followup_suggestions: 2,
    allow_analyst_mode: true, require_citation: true,
    generation: defaultGenerationOptions(),
    retrieval: defaultRetrievalRuntimeOptions(),
    runtime: defaultAgentRuntimeOptions(),
  };
}

export interface AgentRuntimeOptions {
  hyde_enabled: boolean; max_react_steps: number; max_queries_per_step: number;
  max_history_turns: number; max_history_chars: number; max_context_chars: number;
  allow_verifier_correction: boolean;
}
export function defaultAgentRuntimeOptions(): AgentRuntimeOptions {
  return {
    hyde_enabled: true, max_react_steps: 4, max_queries_per_step: 4,
    max_history_turns: 12, max_history_chars: 24_000, max_context_chars: 30_000,
    allow_verifier_correction: true,
  };
}

export interface GenerationOptions { model: string; temperature: number; max_output_tokens: number; }
export function defaultGenerationOptions(): GenerationOptions {
  return { model: 'qwen-turbo', temperature: 0.2, max_output_tokens: 1200 };
}

export interface RetrievalRuntimeOptions {
  dense_top_k: number; bm25_top_k: number; rrf_top_k: number; rerank_top_k: number;
  rerank_enabled: boolean;
}
export function defaultRetrievalRuntimeOptions(): RetrievalRuntimeOptions {
  return { dense_top_k: 100, bm25_top_k: 100, rrf_top_k: 20, rerank_top_k: 5, rerank_enabled: true };
}

export interface CitationOutput {
  index: number; chunk_id: string; doc_id: string; doc_title: string; page_range: number[];
  quote: string; score: number; source_status: string; anchor?: CitationAnchor | null;
}

export type AnswerStreamItem =
  | { type: 'delta'; text: string }
  | { type: 'replace'; text: string }
  | { type: 'citation'; citation: CitationOutput }
  | { type: 'completed'; confidence: Confidence; usage: Usage | null }
  | { type: 'failed'; code: string; message: string };

export interface AgentRun {
  assistant_message_id: string; mode: AgentMode;
  rewritten_query: string | null; retrieval_plan: RetrievalPlan;
  retrieval_traces: RetrievalTrace[];
  /** 移植自 Rust AnswerStream（tokio mpsc::UnboundedReceiver） */
  answerStream: AsyncGenerator<AnswerStreamItem>;
  trace: AgentTrace;
  no_answer_reason: import('./index.ts').NoAnswerReason | null;
}

export interface AgentTrace {
  mode: AgentMode; mode_reason: string; rewritten_query: string | null; keywords: string[];
  resolved_refs: ResolvedRef[]; retrieval_plan: RetrievalPlan; prompt_versions: PromptVersions;
  model: string; usage: Usage | null; started_at: string;
  memory_summary?: string; react_steps?: ReactStepTrace[]; stop_reason?: string;
  runtime_components?: RuntimeComponents; cache_key?: string | null;
}

export interface RuntimeComponents {
  reasoner?: string; retriever?: string; reranker?: string; verifier?: string;
}
export function defaultRuntimeComponents(): RuntimeComponents {
  return { reasoner: '', retriever: '', reranker: '', verifier: '' };
}

export interface ReactStepTrace {
  step: number; action: string; decision_summary: string; output?: string | null;
  tool_calls?: ReactToolCallTrace[]; queries?: string[]; rerank_query?: string | null;
  hypothetical_answer?: string | null; retrieved_chunk_ids?: string[]; accepted_chunk_ids?: string[];
  warnings?: string[]; started_at: string; completed_at: string;
}

export interface ReactToolCallTrace {
  id: string; name: string; arguments: unknown; status: string;
  result?: unknown; error?: unknown;
  started_at: string; completed_at: string; duration_ms: number;
}

export interface PromptVersions { persona: string; guardrail: string; mode: string; task: string; }

export interface RewriteOutput {
  rewritten_query: string; keywords: string[]; hypothetical_answer: string | null;
  resolved_refs: ResolvedRef[]; added_constraints: string[]; removed_constraints: string[];
  needs_clarification: boolean; clarification_question: string | null;
}
